# DS Lab Assignment 3 - Python Basics MCQ

**Name**: Debarun Ghosh \
**Student ID**: 25MDA294 \
**Date**: 07 May 2026

**Q1**.  Consider the following code and choose the odd one out
```python
for i in range(1,5): 
print(i**2)
```
a) 16 
b) 25 
c) 1  
d) 4 
e) 9 

In [1]:
#output of q1:
for i in range(1,5):
    print(i**2)

1
4
9
16


**Output** of q1:

The odd one out of 1, 4, 9, 16, 25 is **25** (option **b**)

**Q2**. Prime number are the unique set of numbers widely used in 
mathematics. It is  because of the fact that the number is 
divisible by 1 and itself. Now construct a function that 
determines the whether the given is prime or not and further 
reuse this function to print all the prime number between 1 
and the N(like if N=10 prime numbers are 2,3,5,7 )

In [ ]:
#solve q2
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

def print_primes(n):
    for i in range(2, n + 1):
        if is_prime(i):
            print(i)
# Example usage:
N = 10
print(f"Prime numbers between 1 and {N}:")
print_primes(N)

Prime numbers between 1 and 10:
2
3
5
7


**Q3**. What is the output of the following code 

x = [1, 2, 3, 4, 5] 

print(x[1:4])

a) [1, 2, 3] 
b) [2, 3, 4] 
c) [1, 2, 3, 4] 
d) [2, 3, 4, 5]

In [6]:
x = [1, 2, 3, 4, 5]
print(x[1:4])

[2, 3, 4]


**Output** of q3 is opion **b** --> [2, 3, 4]

**Q4**. Which of the following is TypeError.?
a = “5” 
b = 2 
a) a*b 
b) int(a) + b 
c) a + str(b) 
d) a + b 

In [7]:
#solve q4
a = "5"
b = 2
print(a * b)  # This will print "5555555555" because string multiplication repeats the string
print(int(a) + b)  # This will print 7 because int(a) converts the string "5" to the integer 5
print(a + str(b))  # This will print "52" because str(b) converts the integer 2 to the string "2" and then concatenates it with "5"
print(a + b)  # This will raise a TypeError because you cannot concatenate a string and an integer directly without conversion.

55
7
52


TypeError: can only concatenate str (not "int") to str

**Output** of q4 is option **d** --> a + b

**Q5**.  What does the below code print.?
for i in range(3): 
    pass 
    print(i) 

a) Nothing – i is undefined 
b) 2 
c) 3 
d) 0

In [12]:
for i in range(3):
    pass
print(i)

2


**Q6**. What would be the output data types?
```python
print(3/2)
print(3//2)
```
a) int, int 
b) float, int 
c) float, float 
d) int, float

In [13]:
print(type(3/2))   # 3/2 uses true division → float
print(type(3//2))  # 3//2 uses floor division → int
print(3/2)
print(3//2)

<class 'float'>
<class 'int'>
1.5
1


**Output** of Q6 is option **b** → `float, int`

- `3/2` uses **true division** → result is `1.5` (float)
- `3//2` uses **floor division** → result is `1` (int)

**Q7**. What would be the output of the code below?
```python
x=10
def foo():
    x= 20
    print(x)
foo()
print(x)
```
a) 20,20 
b) 10,10 
c) 20,10 
d) error

In [14]:
x = 10
def foo():
    x = 20        # local variable inside foo, does NOT affect global x
    print(x)
foo()
print(x)

20
10


**Output** of Q7 is option **c** → `20, 10`

- Inside `foo()`, `x = 20` creates a **local** variable — it does not modify the global `x`
- `print(x)` inside `foo()` prints `20` (local scope)
- `print(x)` outside `foo()` prints `10` (global scope, unchanged)

**Q8**. What will be the output?
```python
print(bool(0), bool(''), bool([]))
print(bool(-1), bool('0'), bool([0]))
```
a) False, True 
b) False, True, False, True, False, True 
c) False, False, False, True, True, True 
d) True, False, False, False, True, True

In [15]:
print(bool(0), bool(''), bool([]))
print(bool(-1), bool('0'), bool([0]))

False False False
True True True


**Output** of Q8 is option **c** → `False False False` / `True True True`

- `bool(0)` → **False** (zero is falsy)
- `bool('')` → **False** (empty string is falsy)
- `bool([])` → **False** (empty list is falsy)
- `bool(-1)` → **True** (any non-zero number is truthy)
- `bool('0')` → **True** (non-empty string is truthy, even if content is '0')
- `bool([0])` → **True** (non-empty list is truthy)

**Q9**. Consider the following code and identify the difference between `//` and `%` operator. What is the correct output?
```python
print(3//2)
print(3%2)
print(7//3)
print(7%3)
```
a) 1 1 2 1 
b) 1 1 1 1 
c) 1 2 1 2 
d) 2 1 1 2

In [16]:
print(3//2)   # floor division → quotient = 1
print(3%2)    # modulo → remainder = 1
print(7//3)   # floor division → quotient = 2
print(7%3)    # modulo → remainder = 1

1
1
2
1


**Output** of Q9 is option **a** → `1 1 2 1`

- `//` (**floor division**) returns the integer quotient: `3//2 = 1`, `7//3 = 2`
- `%` (**modulo**) returns the remainder: `3%2 = 1`, `7%3 = 1`

**Q10**. You are given a dataset containing student information in a raw and inconsistent format. Your task is to clean the data and perform basic analysis.

```python
raw_data = [
    "Rahul,78,20",
    "Anita,85,21",
    "John,NA,19",
    "Meena,92,20",
    "Kiran,,22",
    "Arjun,67,twenty",
    "Divya,88,21"
]
```

Tasks:
1. **Data Cleaning** — remove invalid records (missing values, 'NA', non-numeric marks or age); convert marks and age to int
2. Calculate **Average Marks** after cleaning
3. Display names of students whose **marks are even**
4. Display the name of the **topper**
5. Allocate **batches** (size ≥ 2) and display

In [17]:
raw_data = [
    "Rahul,78,20",
    "Anita,85,21",
    "John,NA,19",
    "Meena,92,20",
    "Kiran,,22",
    "Arjun,67,twenty",
    "Divya,88,21"
]

# 1. Data Cleaning
cleaned = []
for entry in raw_data:
    entry = entry.strip().strip('"').strip(',')
    parts = entry.split(',')
    if len(parts) != 3:
        continue
    name, marks, age = parts
    try:
        marks = int(marks)
        age   = int(age)
        cleaned.append({'name': name, 'marks': marks, 'age': age})
    except ValueError:
        pass   # skip records with NA or non-numeric values

print('Cleaned data:')
for s in cleaned:
    print(f"  {s['name']}: marks={s['marks']}, age={s['age']}")

# 2. Average Marks
avg = sum(s['marks'] for s in cleaned) / len(cleaned)
print(f'\nAverage Marks: {avg:.2f}')

# 3. Students with even marks
print('\nStudents with even marks:')
for s in cleaned:
    if s['marks'] % 2 == 0:
        print(f"  {s['name']} ({s['marks']})")

# 4. Topper
topper = max(cleaned, key=lambda s: s['marks'])
print(f"\nTopper: {topper['name']} ({topper['marks']})")

# 5. Batch allocation (batch size = 2)
batch_size = 2
print('\nBatch Allocation:')
for i, s in enumerate(cleaned):
    batch_num = i // batch_size
    print(f"  {s['name']} → Batch {batch_num}")

Cleaned data:
  Rahul: marks=78, age=20
  Anita: marks=85, age=21
  Meena: marks=92, age=20
  Divya: marks=88, age=21

Average Marks: 85.75

Students with even marks:
  Rahul (78)
  Meena (92)
  Divya (88)

Topper: Meena (92)

Batch Allocation:
  Rahul → Batch 0
  Anita → Batch 0
  Meena → Batch 1
  Divya → Batch 1


**Output** of Q10:

1. **Invalid records removed**: John (NA marks), Kiran (missing marks), Arjun (non-numeric age)
1a. **Cleaned students**: Rahul(78), Anita(85), Meena(92), Divya(88)
2. **Average Marks**: (78+85+92+88) / 4 = **85.75**
3. **Even marks**: Rahul (78), Meena (92), Divya (88)
4. **Topper**: Meena (92)
5. **Batches**: Batch 0 → Rahul, Anita | Batch 1 → Meena, Divya